src/Microsoft.ML.TorchSharp/NasBert/TextClassificationTrainer.cs

```diff
@@ -0,0 +1,857 @@
+            this(env, new Options
+            {
+                OutputColumnName = outputColumnName,
+                Sentence1ColumnName = sentence1ColumnName,
+                Sentence2ColumnName = sentence2ColumnName,
+                LabelColumnName = labelColumnName,
+                BatchSize = batchSize,
+                NumberOfClasses = numberOfClasses,
+                MaxEpoch = maxEpochs,
+                MaxUpdate = maxUpdates,
+                ValidationSet = validationSet,
+            })
+        {
+        }
+
+        internal TextClassificationTrainer(IHostEnvironment env, Options options)
+        {
+            _host = Contracts.CheckRef(env, nameof(env)).Register(nameof(TextClassificationTrainer));
+            _options = options;
+        }
+
+        public SentenceClassificationTransformer Fit(IDataView input)
+        {
+            using (var ch = _host.Start("TrainModel"))
+            using (var pch = _host.StartProgressChannel("Training model"))
+            {
+                var header = new ProgressHeader(new[] { "Total Terms" }, null);
+                var trainer = new Trainer(this);
+                pch.SetHeader(header, e => e.SetMetric(0, trainer.Accuracy));
+                for (int i = 0; i < _options.MaxEpoch && trainer.Updates < _options.MaxUpdate; i++)
+                {
+                    ch.Trace($"Starting epoch {i}");
+                    trainer.Train(input);
+                    ch.Trace($"Finished epoch {i}");
+                    if (_options.ValidationSet != null)
+                        trainer.Validate(pch, ch, i);
+                }
+                _transformer = new SentenceClassificationTransformer(_host, _options, trainer.Model, trainer.Tokenizer.Vocabulary);
+
+                _transformer.GetOutputSchema(input.Schema);
+            }
+            return _transformer;
+        }
+
+        private class Trainer
+        {
+            public BpeTokenizer Tokenizer;
+            public SentenceClassificationModel Model;
+            public torch.Device Device;
+            public BaseOptimizer Optimizer;
+            public optim.lr_scheduler.LRScheduler LearningRateScheduler;
+            private readonly TextClassificationTrainer _parent;
+            public int Updates;
+            public float Accuracy;
+
+            public Trainer(TextClassificationTrainer parent)
+            {
+                _parent = parent;
+                Updates = 0;
+                Accuracy = 0;
+
+                // Get the tokenizer
+                Tokenizer = BpeTokenizer.GetInstance();
+
+                // Initialize the vocab
+                var vocabulary = Tokenizer.Vocabulary;
+                vocabulary.AddMaskSymbol();
+
+                // Initialize the model and load pre-trained weights
+                Model = new SentenceClassificationModel(_parent._options, vocabulary, _parent._options.NumberOfClasses);
```

Comment:
>vocabulary [](http://example.com/codeflow?start=74&length=10)

interesting the model depends on the vocabulary directly and not on the tokenizer.

Reply:
It only depends on it to get a couple of parameters out of it (PadIndex and Count). We should be able to swap that as needed when your new tokenizer is done.